src/testFixtures/groovy/grails/plugin/geb/WebDriverContainerHolder.groovy

```diff
@@ -0,0 +1,143 @@
+
+    RecordingSettings recordingSettings
+    WebDriverContainerConfiguration configuration
+    BrowserWebDriverContainer currentContainer
+
+    WebDriverContainerHolder(RecordingSettings recordingSettings) {
+        this.recordingSettings = recordingSettings
+    }
+
+    boolean isInitialized() {
+        currentContainer != null
+    }
+
+    void stop() {
+        currentContainer?.stop()
+        currentContainer = null
+        configuration = null
+    }
+
+    boolean matchesCurrentContainerConfiguration(WebDriverContainerConfiguration specConfiguration) {
+        specConfiguration == configuration
+    }
+
+    private static int getPort(IMethodInvocation invocation) {
+        try {
+            return (int) invocation.instance.metaClass.getProperty(invocation.instance, 'serverPort')
+        } catch (Exception ignore) {
+            throw new IllegalStateException('Test class must be annotated with @Integration for serverPort to be injected')
+        }
+    }
+
+    boolean reinitialize(IMethodInvocation invocation) {
+        WebDriverContainerConfiguration specConfiguration = new WebDriverContainerConfiguration(
+                getPort(invocation),
+                invocation.getSpec()
+        )
+        if (matchesCurrentContainerConfiguration(specConfiguration)) {
+            return false
+        }
+
+        if (initialized) {
+            stop()
+        }
+
+        configuration = specConfiguration
+        currentContainer = new BrowserWebDriverContainer()
+        Testcontainers.exposeHostPorts(configuration.port)
+        currentContainer.tap {
+            addExposedPort(configuration.port)
+            withAccessToHost(true)
+            start()
+        }
+        if (hostnameChanged) {
+            currentContainer.execInContainer('/bin/sh', '-c', "echo '$hostIp\t${configuration.hostName}' | sudo tee -a /etc/hosts")
+        }
+
+        if (recordingSettings.recordingEnabled) {
+            currentContainer = currentContainer.withRecordingMode(
+                    recordingSettings.recordingMode,
+                    recordingSettings.recordingDirectory,
+                    recordingSettings.recordingFormat
+            )
+        }
+
+        WebDriver driver = new RemoteWebDriver(currentContainer.seleniumAddress, new ChromeOptions())
+        driver.manage().timeouts().implicitlyWait(Duration.ofSeconds(30))
+
+        // Update the browser to use this container
+        Browser browser = (invocation.sharedInstance as ContainerGebSpec).browser
+        browser.driver = driver
+        browser.baseUrl = "${configuration.protocol}://${configuration.hostName}:${configuration.port}"
+
+        return true
+    }
+
+    private boolean getHostnameChanged() {
+        configuration.hostName != ContainerGebConfiguration.DEFAULT_HOSTNAME_FROM_CONTAINER
+    }
+
+    private String getHostIp() {
+        currentContainer.getContainerInfo().getNetworkSettings().getNetworks().entrySet().first().value.ipAddress
```

Comment:
This does not work, as it gets the webdriver container, not the SSH (PortForwardingContainer) containers ipAddress. I have suggested a change in a PR to @jdaugherty's branch.